applications/FluidDynamicsApplication/custom_processes/two_fluids_inlet_process.cpp

```diff
@@ -0,0 +1,221 @@
+//
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "utilities/openmp_utils.h"
+#include "utilities/body_distance_calculation_utils.h"
+#include "utilities/parallel_levelset_distance_calculator.h"
+#include "utilities/variable_utils.h"
+#include "includes/deprecated_variables.h"
+
+// Application includes
+#include "two_fluids_inlet_process.h"
+#include "fluid_dynamics_application_variables.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+
+TwoFluidsInletProcess::TwoFluidsInletProcess(
+    ModelPart& rModelPart,
+    Parameters& rParameters)
+    : Process(), mrInletModelPart(rModelPart) {
+
+    KRATOS_CHECK_VARIABLE_KEY( DISTANCE )
+    KRATOS_CHECK_VARIABLE_KEY( AUX_DISTANCE )
+    KRATOS_CHECK_VARIABLE_KEY( NODAL_AREA )
+
+    // finding the complete model of the inlet model part
+    ModelPart& rRootModelPart = mrInletModelPart.GetRootModelPart();
+
+    // setting the parameters to the private data members of the class
+    mInterfaceNormal = rParameters["two_fluid_settings"]["interface_normal"].GetVector();
+    mInterfacePoint = rParameters["two_fluid_settings"]["point_on_interface"].GetVector();
+    mInletRadius = rParameters["two_fluid_settings"]["inlet_transition_radius"].GetDouble();
+
+    // setting flags for the inlet on nodes and conditions
+    // #pragma omp parallel for
+    // for (int i_node = 0; i_node < static_cast<int>(mrInletModelPart.NumberOfNodes()); ++i_node){
+    for (int i_node = 0; i_node < static_cast<int>( mrInletModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_nodes = mrInletModelPart.NodesBegin() + i_node;
+        it_nodes->Set( INLET, true );
+    }
+    // #pragma omp parallel for
+    for (int i_cond = 0; i_cond < static_cast<int>( mrInletModelPart.NumberOfConditions() ); ++i_cond){
+        // iteration over all conditions
+        auto it_cond = mrInletModelPart.ConditionsBegin() + i_cond;
+        it_cond->Set( INLET, true );
+    }
+
+    // Comment: The historical DISTANCE variable is used to compute a distance field that is then stored in a non-historical variable AUX_DISTANCE
+    //          The functions for the distance computation do not work on non-histirical variables - this is reason for the following procedure (*)
+
+    // (*) temporally storing the distance field as an older version of itself (it can be assured that nothing is over-written at the start)
+    // #pragma omp parallel for
+    for (int i_node = 0; i_node < static_cast<int>( rRootModelPart.NumberOfNodes() ); ++i_node){
+        // iteration over all nodes
+        auto it_node = rRootModelPart.NodesBegin() + i_node;
+        it_node->GetSolutionStepValue(DISTANCE, 2) = it_node->GetSolutionStepValue(DISTANCE, 0);
+    }
+
+    // Preparing the distance computation
+    VariableUtils var_utils;
+    var_utils.SetNonHistoricalVariable( IS_VISITED, 0.0, rRootModelPart.Nodes() );
```

Comment:
there is a flag VISITED (which occupies no memory).

can u use that?

Reply:
I am afraid it won´t work here. The function implementation of `ParallelDistanceCalculator::CalculateDistancesLagrangianSurface` directly use the (deprecated) double variable `IS_VISITED`. A change towards the discussed flag `VISITED` would require extensive changes in the distance calculation functions.

Reply:
uff, can u open an issue about this? btw, why don't you call the VariationalDistanceCalculator? it should be better

Reply:
I am sorry for the delayed answer. I tried a lot to solve this issue but finally did not succeed.

The reason for the usage of this function was its description as 

> A simplified version of CalculateDistances to be used when the rDistanceVar == 0 surface is described by a set of nodes

which avoids the definition of surfaces. Also, the functionality is available for OpenMP and MPI applications.

The main problem I experienced while trying to resort to `VariationalDistanceCalculationProcess` and `TrilinosVariationalDistanceCalculationProcess` is that both depend on the _linear solver_. At this point I get stuck because

- as for as I could see, there is now way to retrieve the already constructed linear solver.
- constructing a new instance of the linear solver (which is probably not optimal) requires the specification of many template arguments. Together with the template arguments of the distance calculation process, I could not even achieve a compiling version.

I am sorry, but after all, I am afraid that I will have to open an issue or take care of the adaptation of the class `CalculateDistancesLagrangianSurface` to work with the `VISITED` flag.

Reply:
UPDATE: After experimenting a lot to finally solve the discussed problem, two approaches (that seemed promising in the beginning) were tested and failed:

- Construction of `VariationalDistanceCalculationProcess` (VDCP) in C++ file of inlet process:
It was possible to construct the serial version of the VDCP. However, it does not seem to be possible to construct its MPI equivalent. The include path settings do not allow to include a header from "trilinos_application", where also the process itself is declared in `Kratos/applications/trilinos_application/custom_processes/trilinos_variational_distance_calculation_process.h.` This makes sense, because otherwise the FluidDynamicsApplication could not be compiled without the TrilinosApplication. Thus, I do not think that altering the include path is possible.

- Construction of both LinearSolver and VDCP in Python:
The idea was to use the simpler constructors in Python and to hand over the complete VDCP or TrilinosVDCP to C++ as an argument of `Process::Pointer`. Simply executing the process from inside the C++ function generated a nice distance field. However, the construction of the process in Python caused far reaching other problems and interfered with steps in the sequence defined by `AnalysisStage`.

@RiccardoRossi @rubenzorrilla, I am extremely sorry but after investing a lot of effort I am afraid that I still need to give up, here. Do you have any idea to solve the issue?

Reply:
Concerning the first approach, it can be solved by creating a Trilinos cpp inlet process, so probably it isn't worth it to take this one.

For the second approach, which is the interference you are talking about?

Reply:
Second approach: The observation is that many variables that are still needed are given the value 0.0. The first of them is e.g. that density and viscosity are set to 0.0 which already causes an error in _set_physical_properties. After setting a control point, it turned out that the `clear()` function is called.

I just looked through the VariationalDistanceProcess itself, once again. It seems as if a communicator was involved. Do you think I could use the "normal" VariationalDistanceProcess also for an MPI application? I did not really consider this so far.

Reply:
Now I see... We need to think in a workaround to circumvent this. Concerning the MPI stuff, we need to use the Trilinos version, since it is constructed with the TrilinosSpace. The serial version has the communicator calls to avoid reimplementing everything for the MPI (the serial communicator calls are somehow "fake" calls).

Reply:
The communicator think would be ideal if it works... If we are lucky, you should be able to do everything you need through the interface of the ModelPart Communicator (`ModelPart.GetCommunicator()`), which should be properly set up in MPI runs.

If it turns out that you actually need to call Trilinos things, it is a bit of a mess right now (we are in the process of cleaning it up). The quick-and-dirty solution would be to use the fact that right now the trilinos application still requires (and links to) the FluidDynamicsApplication. If the class you want to use from the FluidDynamicsApplication is header-only, you should be able to use it (and to derive from it) within the trilinos application using a relative include path (see the trilinos_stokes_initialization_process in the trilinos application). This  is not ideal, but since the clean way of doing it is not ready yet, it could work as a stopgap.

Reply:
A (small) modification to the VariationalDistanceCalculationProcess is suggested in #3954. This is supposed to prepare for the second approach: 

> Construction of both LinearSolver and VDCP in Python